Firebase/Utilities/Logger/GULLogger.m

```diff
@@ -0,0 +1,273 @@
+
+// Allow clients to register their version to include in the log.
+static const char *sVersion = "";
+
+static GULLoggerService kGULLoggerLogger = @"[GULLogger]";
+
+#ifdef DEBUG
+/// The regex pattern for the message code.
+static NSString *const kMessageCodePattern = @"^I-[A-Z]{3}[0-9]{6}$";
+static NSRegularExpression *sMessageCodeRegex;
+#endif
+
+void GULLoggerInitializeASL(BOOL disableDebug, BOOL enableDebug, BOOL forceStderr) {
+  dispatch_once(&sGULLoggerOnceToken, ^{
+    NSInteger majorOSVersion = [[GULAppEnvironmentUtil systemVersion] integerValue];
+    uint32_t aslOptions = ASL_OPT_STDERR;
+#if TARGET_OS_SIMULATOR
+    // The iOS 11 simulator doesn't need the ASL_OPT_STDERR flag.
+    if (majorOSVersion >= 11) {
+      aslOptions = 0;
+    }
+#else
+    // Devices running iOS 10 or higher don't need the ASL_OPT_STDERR flag.
+    if (majorOSVersion >= 10) {
+      aslOptions = 0;
+    }
+#endif  // TARGET_OS_SIMULATOR
+
+    // Override the aslOptions to ASL_OPT_STDERR if the override argument is passed in.
+    if (forceStderr) {
+      aslOptions = ASL_OPT_STDERR;
+    }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"  // asl is deprecated
+    // Initialize the ASL client handle.
+    sGULLoggerClient = asl_open(NULL, kGULLoggerASLClientFacilityName, aslOptions);
+
+    // Set the filter used by system/device log. Initialize in default mode.
+    asl_set_filter(sGULLoggerClient, ASL_FILTER_MASK_UPTO(ASL_LEVEL_NOTICE));
+    sGULLoggerDebugMode = NO;
+    sGULAnalyticsDebugMode = NO;
+    sGULLoggerMaximumLevel = GULLoggerLevelNotice;
+
+    // Use the standard NSUserDefaults if it hasn't been explicitly set.
+    if (sGULLoggerUserDefaults == nil) {
+      sGULLoggerUserDefaults = [NSUserDefaults standardUserDefaults];
+    }
+    BOOL debugMode = [sGULLoggerUserDefaults boolForKey:kGULPersistedDebugModeKey];
+
+    if (disableDebug) {  // Default mode
+      [sGULLoggerUserDefaults removeObjectForKey:kGULPersistedDebugModeKey];
+    } else if (enableDebug || debugMode) {  // Debug mode
+      [sGULLoggerUserDefaults setBool:YES forKey:kGULPersistedDebugModeKey];
+      asl_set_filter(sGULLoggerClient, ASL_FILTER_MASK_UPTO(ASL_LEVEL_DEBUG));
+      sGULLoggerDebugMode = YES;
+    }
+
+    // We should disable debug mode if we are running from App Store.
+    if (sGULLoggerDebugMode && [GULAppEnvironmentUtil isFromAppStore]) {
+      sGULLoggerDebugMode = NO;
+    }
+
+    sGULClientQueue = dispatch_queue_create("GULLoggingClientQueue", DISPATCH_QUEUE_SERIAL);
+    dispatch_set_target_queue(sGULClientQueue,
+                              dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0));
+
+#ifdef DEBUG
+    sMessageCodeRegex =
+        [NSRegularExpression regularExpressionWithPattern:kMessageCodePattern options:0 error:NULL];
+#endif
+  });
+}
+
+void GULSetAnalyticsDebugMode(BOOL analyticsDebugMode) {
+  GULLoggerInitializeASL(NO, NO, NO);
```

Comment:
This is now functionally different than what came before.

Previously if your first call was FIRSetAnalyticsDebugMode we would initialize FIRLogger, *including evaluating the command-line arguments*. This now disregards the command-line arguments.